os/kernel/task/task_event_notify.c

```diff
@@ -0,0 +1,73 @@
+/****************************************************************************
+ *
+ * Copyright 2020 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <arch/irq.h>
+#include <tinyara/arch.h>
+#include <tinyara/sched.h>
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: task_event_notify
+ *
+ * Description:
+ *   This function sends an event to specific pid task.
+ *   If that task was in blocked queue, it will be unblocked when receiving event.
+ *
+ * Inputs:
+ *   pid - The task ID of the task to receive the event.
+ *   data - The data which passed to user.
+ *
+ * Return Value:
+ *   OK on success; or ERROR on failure
+ *
+ ****************************************************************************/
+
+int task_event_notify(pid_t pid, void *data)
+{
+	int saved_state;
+	struct tcb_s *tcb = sched_gettcb(pid);
+	if (tcb == NULL) {
+		return ERROR;
+	}
+
+	saved_state = irqsave();
+
+	if (tcb->event_data[0] == NO_PENDING_DATA) {
+		/* If this is the first event, save the date in the first. */
+		tcb->event_data[0] = data;
+	} else {
+		/* If there is already pending event, queueing new data. */
+		tcb->queueing = true;
+		tcb->event_data[1] = data;
+	}
```

Comment:
Minor one,
In my opinion,  it is same but ((void*)0) or NULL is more practical value instead of 0 for NO_PENDING_DATA.
also, 'tcb.queuing is unnecessary, it always set when NO_PENDING_DATA is true.
BTW, if I were you I will divide array to understand more easier if we can assume that we will not expand structure in future...of course if we consider queue based on pipeline buffer in future, it will be better.
```
if (tcb->act_event_data == NULL) {
    tcb->act_event_data = data;
} else {
   tcb->pending_event_data = data;
}
...
if (tcb->act_event_data != NULL) {
    ret = tcb->act_event_data;
    tcb->act_event_data = tcb->act_pend_data; //acceptable even if it is null
}
```

Reply:
Thanks for review.
1. Now I'm worried that sometime we want to pass the real '0' value. So I'm not sure how to define the NO_PENDING_DATA. In this case, we should change the value to something. So I will define NO_PENDING_DATA as ((void *)0).
2. Yes. If we don't have plan for expand, your suggestion is better. I will consider more. Thanks.

Reply:
Then I suggest additional struct below if this can be added to somewhere in os/include.
```
struct event_data {
    void *data;
    ssize_t size;
}
```
Then you don't have to use 'queueing' and it is more practical code, In my opinion.
Also receiver can use malloc properly with given parameter.

Reply:
I will consider your suggestion with another commit/PR.
Because now I want to use this from kernel to user, then there should be some logic for copying between kernel and user. I'm not sure that can have enough performance.